recipe/recipe.go

```diff
@@ -36,3 +42,335 @@ type instance struct {
+				}
+			}
+			if len(instanceInvalidFields) > 0 {
+				for _, iField := range instanceInvalidFields {
+					instanceInvalidFields[i] = "output-instances[" + strconv.Itoa(i) + "]." + iField
+				}
+			}
+			missingFields = append(missingFields, instanceMissingFields...)
+			invalidFields = append(invalidFields, instanceInvalidFields...)
+		}
+	} else {
+		missingFields = append(missingFields, "output-instances")
+	}
+
+	if missingFields != nil {
+		return fmt.Errorf("missing mandatory fields: %v", missingFields)
+	}
+
+	if invalidFields != nil {
+		return fmt.Errorf("invalid fields: %v", invalidFields)
+	}
+
+	return nil
+
+}
+
+//ValidateAddInstance - checks if fields of OutputInstances are not empty
+func (instance *Instance) ValidateAddInstance() error {
+	var missingFields []string
+	var invalidFields []string
+
+	instanceMissingField, instanceInvalidField := instance.validateInstance()
+	missingFields = append(missingFields, instanceMissingField...)
+	invalidFields = append(invalidFields, instanceInvalidField...)
+
+	if missingFields != nil {
+		return fmt.Errorf("missing mandatory fields: %v", missingFields)
+	}
+
+	if invalidFields != nil {
+		return fmt.Errorf("invalid fields: %v", invalidFields)
+	}
+
+	return nil
+}
+
+//ValidateAddCodelist - checks if fields of Codelist are not empty
+func (codelist *CodeList) ValidateAddCodelist() error {
+	var missingFields []string
+	var invalidFields []string
+
+	codelistMissingFields, codelistinvalidFields := codelist.validateCodelists()
+	missingFields = append(missingFields, codelistMissingFields...)
+	invalidFields = append(invalidFields, codelistinvalidFields...)
+
+	if missingFields != nil {
+		return fmt.Errorf("missing mandatory fields: %v", missingFields)
+	}
+
+	if invalidFields != nil {
+		return fmt.Errorf("invalid fields: %v", invalidFields)
+	}
+
+	return nil
+}
+
+//ValidateUpdateRecipe - checks updates of recipe for PUT request
+func (recipe *Response) ValidateUpdateRecipe() error {
+	var missingFields []string
+	var invalidFields []string
+	validFormats := map[string]bool{
+		"v4": true,
+	}
+
+	// Validation to check at least one field of the recipe is updated
+	allStringFieldsEmpty := recipe.ID == "" && recipe.Format == "" && recipe.Alias == ""
+	allRemainingFieldsNil := recipe.InputFiles == nil && recipe.OutputInstances == nil
+	if allStringFieldsEmpty && allRemainingFieldsNil {
+		invalidFields = append(invalidFields, "no recipe fields updates given")
+	}
+
+	if recipe.ID != "" {
+		invalidFields = append(invalidFields, "id cannot be changed")
+	}
+	if recipe.Format != "" {
+		if !validFormats[recipe.Format] {
+			invalidFields = append(invalidFields, "format is not valid")
+		}
+	}
+
+	if recipe.InputFiles != nil && len(recipe.InputFiles) > 0 {
+		for i, file := range recipe.InputFiles {
+			if file.Description == "" {
+				missingFields = append(missingFields, "input-files["+strconv.Itoa(i)+"].description")
```

Comment:
I feel like in this case, during an update, providing an empty string is more _invalid_ than it is missing

Reply:
Done in commit "optimising code by adding sub-functions and global var"